### DISCUSSION TITLE
Truth Table Format

### DISCUSSION THREAD

**Discussion by ddotcole** [hneemann/Digital#644] [General]
Hello,

I am looking at importing a truth table into Digital created with a spreadsheet.  I see when a truth table is up, one can open a file of the format "Truth Table".  What is this format and is there an easy way to make and import truth tables into Digital?

Thanks,

David Cole

**Comment by hneemann** [comment]
The file format is a simple XML format. The result columns are stored as a BASE64 encoded array of bytes, where the bytes can have only three values:

```
0: 0
1: 1
2: don't care
```

Although there is a function in almost every programming language to create BASE64, it is very difficult to create this format by hand.

This is an example of an 3 bit adder:

``` XML
<?xml version="1.0" encoding="utf-8"?>
<truthTable>
  <variables>
    <variable name="A_2"/>
    <variable name="A_1"/>
    <variable name="A_0"/>
    <variable name="B_2"/>
    <variable name="B_1"/>
    <variable name="B_0"/>
  </variables>
  <results>
    <result>
      <name>S_2</name>
      <values class="BoolTable">
        <table>AAAAAAEBAQEAAAABAQEBAAAAAQEBAQAAAAEBAQEAAAABAQEBAAAAAAEBAQAAAAABAQEAAAAAAQEBAAAAAAEBAQ==</table>
      </values>
    </result>
    <result>
      <name>S_1</name>
      <values class="BoolTable">
        <table>AAABAQAAAQEAAQEAAAEBAAEBAAABAQAAAQAAAQEAAAEAAAEBAAABAQABAQAAAQEAAQEAAAEBAAABAAABAQAAAQ==</table>
      </values>
    </result>
    <result>
      <name>S_0</name>
      <values class="BoolTable">
        <table>AAEAAQABAAEBAAEAAQABAAABAAEAAQABAQABAAEAAQAAAQABAAEAAQEAAQABAAEAAAEAAQABAAEBAAEAAQABAA==</table>
      </values>
    </result>
    <result>
      <name>C</name>
      <values class="BoolTable">
        <table>AAAAAAAAAAAAAAAAAAAAAQAAAAAAAAEBAAAAAAABAQEAAAAAAQEBAQAAAAEBAQEBAAABAQEBAQEAAQEBAQEBAQ==</table>
      </values>
    </result>
  </results>
</truthTable>
```

The tag `<table>` contains the result column values.

**Comment by ddotcole** [comment]
Thanks for the reply hneeman.  I'll give it a go with a little programming and a csv file to see if I can create this format.

**Reply by hneemann** [reply]
It is easy to create a new XML marshaller, so i have added one, which creates a more readable output.
Now a .tru xml file can also look like this:
``` xml
<?xml version="1.0" encoding="utf-8"?>
<truthTable>
  <variables>
    <variable name="A_2"/>
    <variable name="A_1"/>
    <variable name="A_0"/>
    <variable name="B_2"/>
    <variable name="B_1"/>
    <variable name="B_0"/>
  </variables>
  <results>
    <result>
      <name>S_2</name>
      <values class="BoolTable">0000111100011110001111000111100011110000111000011100001110000111</values>
    </result>
    <result>
      <name>S_1</name>
      <values class="BoolTable">0011001101100110110011001001100100110011011001101100110010011001</values>
    </result>
    <result>
      <name>S_0</name>
      <values class="BoolTable">0101010110101010010101011010101001010101101010100101010110101010</values>
    </result>
    <result>
      <name>C</name>
      <values class="BoolTable">0000000000000001000000110000011100001111000111110011111101111111</values>
    </result>
  </results>
</truthTable>
```

'0' and '1' means what you expect, any other character means "don't care".
Please try the latest [pre release](https://infdigital.dhbw-mosbach.de/).

**Comment by ddotcole** [comment]
Sweet, I'll give that a try.

**Comment by ddotcole** [comment]
Thanks for the help Helmut.  I downloaded the prerelease and was able to cobble together a program to convert a csv file to the Truth Table format you provided above.  Definitely appreciate the help and enjoy the utility this program provides.

**Comment by xotmatrix** [comment]
Thank you for allowing the table to be encoded with normal text. Manually editing .tru files has been very tedious in the past.

**Reply by hneemann** [reply]
There is also a ticket requesting a CSV import (#631). Maybe I should implement a CSV import/export.
It seems to me that a lot of people would appreciate it.

**Reply by ddotcole** [reply]
I know I would.  I think it would align more with how ROM & RAMs are loaded within the program.

**Reply by hneemann** [reply]
The latest [pre release](https://infdigital.dhbw-mosbach.de/) contains a csv import and export.

**Reply by ddotcole** [reply]
Hi Helmut,

I got a chance to try the new pre-release.  I am able to import CSV as long as the number of outputs equals the number of inputs or less.  If there are more outputs then inputs then an error occurs: "To many values in one line".  This does not happen if I convert the csv to your truth table format.

Let me know if you'd like a copy of my csv & tru table to check against.  I can't seem to post them on here.

**Reply by hneemann** [reply]
Thanks for pointing that out! The latests pre release fixes this issue.

BTW: The CSV Importer has a test coverage of 98% including several negative tests and still such errors remain undetected! 
Software development is a difficult business! :smile:

**Comment by Bob-May** [comment]
I suggest that you add something about the base64 encoding to the documentation.
It was not at all obvious to me how to format the truth table for a LUT component until I found this discussion.

Alternately, consider adding a binary column to the lookup data window.  Why force users to consult an external table?

**Reply by hneemann** [reply]
Please try the latest pre-release version, which no longer uses base64 encoding and also allows importing and exporting CSV files, which are much easier to create.

**Reply by Bob-May** [reply]
Thank you for your prompt reply, and more importantly thank you for creating and maintaining digital.

Can you explain how to define a LUT component's truth table via a csv?  It is nowhere obvious.
![image](https://user-images.githubusercontent.com/47577359/110689967-562e4700-81b1-11eb-8ca9-4607e25f22d8.png)

Would it be possible to define and edit truth tables in a format similar to what is displayed from the Analysis option?  
![image](https://user-images.githubusercontent.com/47577359/110690262-b7eeb100-81b1-11eb-8c18-9c68e53c59b8.png)

**Reply by hneemann** [reply]
Each input of the LUT describes one bit of the address. The inputs are labeled with the bit number. And the corresponding value is then output.
If you want to create a LUT via the truth table dialog, you can define the truth table in this dialog and then create a LUT via Create->Circuit with LUTs.

**Reply by Bob-May** [reply]
Very cool!
